### PR TITLE
Add -u param to usermod in linux_user resource when using non_unique

### PR DESCRIPTION
**Pull request by Ppjet6**
Signed-off-by: Maxime “pep” Buquet <pep@collabora.com>

### Description

`-u` (uid) parameter is required by usermod when using `-o` (non_unique). This patch adds the parameter if needed.

I don't have tests for it as apparently there is no test yet for usermod in the linux_spec. I can try to come up with some tests if somebody quickly guides me through it.

### Check List

- [ ] New functionality includes tests
- [ ] All tests pass
- [ ] RELEASE\_NOTES.md, has been updated if required (not required for bugfixes, required for API changes)
- [X] All commits have been signed-off for the Developer Certificate of Origin. See <https://github.com/chef/chef/blob/master/CONTRIBUTING.md#developer-certification-of-origin-dco>

### CODE REVIEW

File: lib/chef/provider/user/linux.rb [hunk 1]
```diff
@@ -58,6 +58,7 @@ def universal_options
 
         def usermod_options
           opts = []
+          opts << "-u %d" % new_resource.uid if non_unique?
```
Comment (lamont-granquist) [inline lib/chef/provider/user/linux.rb]: you probably want `opts += [ "-u", "%d" ]` here

Comment (lamont-granquist) [inline lib/chef/provider/user/linux.rb]: uh, well, avoid the interpolation entirely:  `opts += [ "-u", new_resource.uid ]`

**Comment by lamont-granquist** [comment]
best place to add a spec would most likely be `spec/functional/resource/user/useradd_spec.rb`

probably does need a spec, since i think the original patch most likely didn't work, or at least needs a func test to validate it does work.

File: lib/chef/provider/user/linux.rb (hunk 1 above)
Comment (Ppjet6) [inline lib/chef/provider/user/linux.rb]: Thanks, I'm new to ruby I basically copied things that were around. Would it be best to add another patch? Rebase?

Comment (lamont-granquist) [inline lib/chef/provider/user/linux.rb]: probably rebase and squash, but either way

**Comment by coderanger** [comment]
I think something is wonky with this PR, unit tests on master are busted with:

```
Failures:

  1) Chef::Provider::User::Linux when setting option when the resource has a different home directory and supports home directory management should set -m -d /homedir

     Failure/Error: opts += [ "-u", new_resource.uid ] if non_unique?

     

     NoMethodError:

       undefined method `non_unique?' for #<Chef::Provider::User::Linux:0x00000023d47e98>

     Shared Example Group: "a useradd-based user provider" called from ./spec/unit/provider/user/linux_spec.rb:40

     # ./lib/chef/provider/user/linux.rb:61:in `usermod_options'

     # ./spec/support/shared/unit/provider/useradd_based_user_provider.rb:115:in `block (4 levels) in <top (required)>'

  2) Chef::Provider::User::Linux when setting option when the resource has a different home directory and supports home directory management (using real attributes) should set -m -d /homedir

     Failure/Error: opts += [ "-u", new_resource.uid ] if non_unique?

     

     NoMethodError:

       undefined method `non_unique?' for #<Chef::Provider::User::Linux:0x00000023d488c0>

     Shared Example Group: "a useradd-based user provider" called from ./spec/unit/provider/user/linux_spec.rb:40

     # ./lib/chef/provider/user/linux.rb:61:in `usermod_options'

     # ./spec/support/shared/unit/provider/useradd_based_user_provider.rb:128:in `block (4 levels) in <top (required)>'

  3) Chef::Provider::User::Linux when managing a user runs usermod with the computed command options

     Failure/Error: opts += [ "-u", new_resource.uid ] if non_unique?

     

     NoMethodError:

       undefined method `non_unique?' for #<Chef::Provider::User::Linux:0x00000023cda4d8>

     Shared Example Group: "a useradd-based user provider" called from ./spec/unit/provider/user/linux_spec.rb:40

     # ./lib/chef/provider/user/linux.rb:61:in `usermod_options'

     # ./lib/chef/provider/user/linux.rb:31:in `manage_user'

     # ./spec/support/shared/unit/provider/useradd_based_user_provider.rb:208:in `block (3 levels) in <top (required)>'

  4) Chef::Provider::User::Linux when managing a user does not set the -r option to usermod

     Failure/Error: opts += [ "-u", new_resource.uid ] if non_unique?

     

     NoMethodError:

       undefined method `non_unique?' for #<Chef::Provider::User::Linux:0x00000023c9dd08>

     Shared Example Group: "a useradd-based user provider" called from ./spec/unit/provider/user/linux_spec.rb:40

     # ./lib/chef/provider/user/linux.rb:61:in `usermod_options'

     # ./lib/chef/provider/user/linux.rb:31:in `manage_user'

     # ./spec/support/shared/unit/provider/useradd_based_user_provider.rb:219:in `block (3 levels) in <top (required)>'

  5) Chef::Provider::User::Linux when managing a user CHEF-3429: does not set -m if we aren't changing the home directory

     Failure/Error: opts += [ "-u", new_resource.uid ] if non_unique?

     

     NoMethodError:

       undefined method `non_unique?' for #<Chef::Provider::User::Linux:0x00000023c6d4c8>

     Shared Example Group: "a useradd-based user provider" called from ./spec/unit/provider/user/linux_spec.rb:40

     # ./lib/chef/provider/user/linux.rb:61:in `usermod_options'

     # ./lib/chef/provider/user/linux.rb:31:in `manage_user'

     # ./spec/support/shared/unit/provider/useradd_based_user_provider.rb:228:in `block (3 levels) in <top (required)>'
```

**Comment by adamleff** [comment]
@coderanger agreed, just stumbled upon the same thing.
